Cython/Compiler/Buffer.py

```diff
@@ -679,8 +679,14 @@ def get_type_information_cname(code, dtype, maxdepth=None):
                      for f in fields]
             typecode.putln("static __Pyx_StructField %s[] = {" % structinfo_name, safe=True)
             for f, typeinfo in zip(fields, types):
+                if dtype.needs_cpp_construction:
+                    # for cpp: offsetof(StrucName, memberName)
+                    struct_type = dtype.cname
```

Comment:
1) This looks to me like an abuse of `needs_cpp_construction` when you're really just trying to detect if we're building in C++. I get why it's convenient, but `needs_cpp_construction` is only really about whether to generate C++ constructor and destructor code which doesn't seem relevant here (i.e. I'd rather not have to buffers just because I've changed something about how destructors are called in a cdef class).

    I think you can get the value of `cpp` from `code.funcstate` or `code.globalstate` somewhere.

2) Is this a problem with older MSVC versions? `offsetof(struct StructName, memberName)` seems fine in recent versions.

Reply:
> 2\. Is this a problem with older MSVC versions?

What version do you use?
Github actions and my local installation detects this as an error.

Reply:
https://godbolt.org/z/anEch1f3K (unless there's some detail that I missed). That lets you go back to MSVC 2015, and it seems fine there. I didn't go further.

Reply:
> 1. (i.e. I'd rather not have to buffers just because I've changed something about how destructors are called in a cdef class

absolutely agree

Reply:
Well, the problem is not really the `offsetof(struct StrucName, memberName)`, but probably the surrounding of this code.
You can look [here](https://github.com/0dminnimda/cython/runs/3214687998#step:5:2416)

Reply:
> > 1. (i.e. I'd rather not have to buffers just because I've changed something about how destructors are called in a cdef class
> 
> absolutely agree

this is the reason why I asked about it [here](https://github.com/cython/cython/pull/4324#discussion_r681300618)
^_^

Reply:
`runtests.py` runs `Microsoft (R) C/C++ Optimizing Compiler Version 19.28.29336 for x64` version

Reply:
Oh, I just forgot, the point is, the generated code for the cpp looks like this:
```cpp
static __Pyx_StructField __Pyx_StructFields_nn_struct____pyx_t_8memslice_MyStruct__const__[] = {
  {&__Pyx_TypeInfo_signed__char, "a", offsetof(struct __pyx_t_8memslice_MyStruct const , a)},
  {&__Pyx_TypeInfo_signed__char, "b", offsetof(struct __pyx_t_8memslice_MyStruct const , b)},
  {&__Pyx_TypeInfo_PY_LONG_LONG, "c", offsetof(struct __pyx_t_8memslice_MyStruct const , c)},
  {&__Pyx_TypeInfo_int, "d", offsetof(struct __pyx_t_8memslice_MyStruct const , d)},
  {&__Pyx_TypeInfo_int, "e", offsetof(struct __pyx_t_8memslice_MyStruct const , e)},
  {NULL, NULL, 0}
};
```

and this `const` spoils everything for msvc

Reply:
AFAIU, adding `const` does not change anything, so the changes did not break anything

Reply:
`"memslice.pyx":1169`
const memslice

Reply:
Should we even have this `const` inside `offsetof`?

Reply:
> 1. I think you can get the value of `cpp` from `code.funcstate` or `code.globalstate` somewhere.

AFAIU, we compile this piece of code only in the specified language, so probably, yes, it will do.

Reply:
The original code was added [here](https://github.com/cython/cython/commit/20a4a188e90cac5fe870cde73ada7b7173a78e55#diff-36a782bc4ad684402e6bf0b6e257873a323407a7b7c80a00b4cbbc19f6cc3a73R1153) and used [`__Pyx_FIELD_OFFSET`](https://github.com/cython/cython/blob/20a4a188e90cac5fe870cde73ada7b7173a78e55/Cython/Compiler/Buffer.py#L716) (now deprecated), which did not cause such errors. A bug was added during the transition from `__Pyx_FIELD_OFFSET` to `offsetof`.

Reply:
But in the end, my few tests with `__Pyx_FIELD_OFFSET`  show that there is no difference between `__Pyx_FIELD_OFFSET(struct StrucName const, memberName)` and `offsetof(struct StrucName, memberName)` for msvc

Reply:
> and `offsetof(struct StrucName, memberName)` for msvc

or `offsetof(StrucName, memberName)`

Reply:
> transition from `__Pyx_FIELD_OFFSET` to `offsetof`

[here](https://github.com/cython/cython/commit/012f19376d97ce0d707e73c315d780260021d922)

Reply:
> Should we even have this const inside offsetof?

I don't think so. You could probably do `if dtype.is_cv_qualified: dtype = dtype.base_type` (and do it whether you're in C or in C++ mode)

Reply:
> You could probably do `if dtype.is_cv_qualified: dtype = dtype.cv_base_type` (and do it whether you're in C or in C++ mode)

Yes, I think it will do

Reply:
Ok, @da-woods, thanks for the help, it worked as expected!